=cdd #18768 Cache serialization of read and write msg

patriknw opened the pull request:
Something I did on the plane last week.

The WriteAggregator and ReadAggregator typically send
the same message to several replicas and by caching the serialized bytes
we avoid doing the same thing for each message

add a test for WriteAggregator

patriknw committed:
=cdd #18768 Cache serialization of read and write msg

The WriteAggregator and ReadAggregator typically send
the same message to several replicas and by caching the serialized bytes
we avoid doing the same thing for each message

and add a test for WriteAggregator

viktorklang commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala:
```diff
@@ -1367,6 +1367,8 @@ private[akka] class WriteAggregator(
     case ReceiveTimeout ⇒ reply(ok = false)
   }
 
+  def senderAddress(): Address = sender().path.address
```
why?

patriknw replied:
needed for the new test `WriteAggregatorSpec`, which is not really related to the cache (I just increased the tests coverage a bit)

viktorklang commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatorMessageSerializer.scala:
```diff
@@ -80,9 +171,9 @@ class ReplicatorMessageSerializer(val system: ExtendedActorSystem)
-    case m: Write          ⇒ writeToProto(m).toByteArray
-    case WriteAck          ⇒ dm.Empty.getDefaultInstance.toByteArray
-    case m: Read           ⇒ readToProto(m).toByteArray
+    case m: Write          ⇒ writeCache.getOrAdd(m, () ⇒ writeToProto(m).toByteArray)
```
Instead of allocating this function on each Write, why not prove the function when constructing the SmallCache?

patriknw replied:
good point, thanks

viktorklang commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatorMessageSerializer.scala:
```diff
@@ -21,12 +21,103 @@ import akka.serialization.BaseSerializer
   extends SerializerWithStringManifest with SerializationSupport with BaseSerializer {
+  import ReplicatorMessageSerializer.SmallCache
+
+  private val readCache = new SmallCache[Read, Array[Byte]](4)
```
What is the foreseen implication of having a non-expiring cache for this?

patriknw replied:
I didn't consider it to be a big deal to keep around a few if no more activity, but I can add simple TTL eviction

viktorklang commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatorMessageSerializer.scala:
```diff
@@ -21,12 +21,103 @@ import akka.serialization.BaseSerializer
+
+    def add(x: (A, B)): Unit = {
+      val i = n.incrementAndGet()
+      elements(i & mask) = x
```
What if `x` is `null`?

patriknw replied:
I'm not using null, but I think it can actually work, because the cache elements are initially `null.

patriknw commented:
@viktorklang I have added the cache eviction and changed so that the `getOrAdd` function is passed in to the constructor instead, see 3302b14. Thanks for feedback!

patriknw commented:
Refs #18768
